### PR TITLE
Decision-26435_Angular_Upgrade

### PR DESCRIPTION
**Pull request by krupasodha**
Angular upgrade version  7.2.3 -> 12.2.5

Upgarded using angular update command  => ng update 
updated some dependencies manually which were not handled by angular update command

angular.json is updated by Angular update command

### CODE REVIEW

File: libs/angular-breadcrumb/src/searchbox/searchbox.component.spec.ts [hunk 1]
```diff
@@ -56,7 +56,7 @@ describe("BreadcrumbRoute Popup Component", () => {
   });
   describe('when there is no input for search data function', () => {
 
-    beforeEach(async(() => {
+    beforeEach(waitForAsync(() => {
```
Comment (swatantrajain-sapiens) [inline libs/angular-breadcrumb/src/searchbox/searchbox.component.spec.ts:59]: why waitForAsync?

File: apps/demo/src/app/demo/breadcrumb-template.component.ts [hunk 2]
```diff
@@ -1,12 +0,0 @@
-import {Component, Input} from "@angular/core";
-@Component({
-  selector: 'dynamic-list-item',
-  template: `
-    <div>Template Item
-    {{item.breadcrumb.label}}
-    </div>
-  `
-})
-export class BreadcrumbTemplate{
```
Comment (swatantrajain-sapiens) [inline apps/demo/src/app/demo/breadcrumb-template.component.ts:10]: Its not relevant anymore?

File: libs/angular-breadcrumb/src/searchbox/searchbox.component.spec.ts (hunk 1 above)
Comment (krupasodha) [reply libs/angular-breadcrumb/src/searchbox/searchbox.component.spec.ts:59]: waitForAysnc is replacement of async provided by Angular. async is deprecated in Angular 12

File: apps/demo/src/app/demo/breadcrumb-template.component.ts (hunk 2 above)
Comment (krupasodha) [reply apps/demo/src/app/demo/breadcrumb-template.component.ts:10]: This looked like dummy class which was not used anywhere and not declared with any module. Angular does not allow to have such components now without declaring it in any module

**Comment by swatantrajain-sapiens** [comment]
LGTM

File: libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts [hunk 3]
```diff
@@ -435,12 +448,15 @@ export function click(el: DebugElement | HTMLElement, eventObj: any = ButtonClic
 //region test components
 @Component({
   template: `
-      <dcn-breadcrumb-popup [breadcrumbDropDown]="testBreadCrumb"></dcn-breadcrumb-popup>
+      <dcn-breadcrumb-popup [breadcrumbDropDown]="testBreadCrumb" [rootActivatedRoute]="rootActivatedRoute"></dcn-breadcrumb-popup>
```
Comment (urielTwito) [inline libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts:451]: why do we need rootActivatedRoute

File: libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts [hunk 4]
```diff
@@ -107,17 +113,20 @@ describe("Breadcrumb Popup Component", () => {
           });
         });
 
-        it('should have text and icon for dynamic links', () => {
+        it('should have text and icon for dynamic links', waitForAsync(() => {
```
Comment (urielTwito) [inline libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts:116]: why not to use fakeAsync?

**Review by urielTwito — COMMENTED** [review]
not required but can we use angular library

File: libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts (hunk 4 above)
Comment (krupasodha) [reply libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts:116]: I think I can try using fakeAsync also, but any reason for preferring fakeAsync over waitForAsync ?

File: libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts (hunk 3 above)
Comment (krupasodha) [reply libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts:451]: rootActivatedRoute was added as an input for breadcrumb-popup.component as a part of some other recent ticket. I added this input in its spec to enhance the spec and cover this newly added input field

File: libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts (hunk 4 above)
Comment (urielTwito) [reply libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts:116]: The tests will run faster (will not wait for all the timeouts in the component)
this is our best practice in NG, all our tests should use fakeAsync not Async.
we do this in every beforeEach  (even if we do not have async operation in the code)

**Comment by krupasodha** [comment]
> not required but can we use angular library

Not sure I understand correctly. What do you mean by using angular library ? 
 we already have this published as angular library with npm

**Comment by urielTwito** [comment]
> not required but can we use angular library

if you will build new angular library with angular cli.
the structure of the the library is different. 
they use  ng-packager in a different way
they use AOT (I am not sure if we use AOT now)

again it is not mandatory, but if it is no a big task maybe it is worth doing this.

File: libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts (hunk 4 above)
Comment (krupasodha) [reply libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts:116]: I will try replacing all waitForAsync with fakeasync wherever possible

**Comment by krupasodha** [comment]
> > not required but can we use angular library
> 
> if you will build new angular library with angular cli. the structure of the the library is different. they use ng-packager in a different way they use AOT (I am not sure if we use AOT now)
> 
> again it is not mandatory, but if it is no a big task maybe it is worth doing this.

I think we are already using AOT for breadcrumb repo now. The description in github page for this repo says :

This project is a fork of the now deprecated https://www.npmjs.com/package/ng2-navigator
It was transitioned based on https://github.com/jasonaden/angular-cli-lib-example to have proper bundling and AoT support

**Comment by urielTwito** [comment]
It is your call.

File: libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts (hunk 4 above)
Comment (krupasodha) [reply libs/angular-breadcrumb/src/popup/breadcrumb-popup.component.spec.ts]: @urielTwito Done. Replaced All WaitForAsync with FakeAsync

**Pull request merged**